compiler/rustc_hir_typeck/src/closure.rs

```diff
@@ -364,36 +347,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let tcx = self.tcx;
 
         let trait_def_id = projection.trait_def_id(tcx);
-
-        let is_fn = tcx.is_fn_trait(trait_def_id);
-
-        let coroutine_trait = tcx.lang_items().coroutine_trait();
-        let is_gen = coroutine_trait == Some(trait_def_id);
-
-        if !is_fn && !is_gen {
-            debug!("not fn or coroutine");
+        // For now, we only do signature deduction based off of the `Fn` traits.
```

Comment:
I don't expect us to extend this to deduce other coroutines' signatures: https://github.com/rust-lang/rust/pull/109338

Or at least, when we *do* start doing so, this method should be reworked to be cleaner and more extensible. Right now there's a lot of funky control flow for absolutely no reason 😅